### ISSUE TITLE
Reduce model graph size

### ISSUE THREAD

**Issue by Shazwazza** [umbraco/Umbraco-CMS#3543]
We would like to help reduce memory footprint by changing our persistence models to not have deep object graphs. This is a major version type of change and cannot be done in a minor version. The first thing we should do is remove this: `IContent.IContentType` and `IContent.ITemplate` reference. We cache a lot of content and for every bit of content we cache, we are caching the entire object graphs of it's corresponding `IContentType` and `ITemplate` :disappointed: and these are *not* cached references, these are cached clones. IContent needs to be changed to `IContent.ContentTypeId` and `IContent.TemplateId`. Ideally we'd pull apart other models too such as `IContentType` itself which has references to other `IContentType`, property types, groups, etc... based on compositions, etc... which means that a single `IContentType` could be caching clones of several other `IContentType` that is already in the cache for itself. There's plenty of examples of this in the core but it's not a super quick fix. IMO the quickest and biggest win is to change the `IContent` model mentioned above.

**Comment by kjac** [comment]
@Shazwazza V8 is indeed an appropriate time to perform this kind of breaking changes. I'm certain there are a LOT of code bases out there that utilize the `IContentType` and `ITemplate` properties on `IContent`.

I'd be happy to help with the refactoring if you want it.

Also... #3544 and #3545 should probably *not* be done in parallel due to the obvious merging conflicts that'll arise.

**Comment by Shazwazza** [comment]
@kjac defo, one thing at a time :) This is only the tip of the iceberg though, there are other models with deep graphs that we shouldn't have but decoupling `IContent` is the best place to start since we have more of these than other models and they get cached along with their graph. I think this work might be scheduled for someone on the core team but we'll be in touch if that doesn't go ahead and you can go for it ! :) //cc @clausjensen

**Comment by clausjensen** [comment]
@Shazwazza yep 😄 we have it scheduled here in Core so for now there's no need for the community to pick up this particular task. (cc @kjac ).

**Comment by kjac** [comment]
Roger 😀

**Comment by bergmania** [comment]
Both child issues are part of the pull request.
